🔧 Enhance Type Safety and Code Quality with PHPStan Level 4 Improvements

anilkumarthakur60 opened the pull request:
This PR upgrades the codebase to **PHPStan level 4**, significantly improving static analysis, type safety, and overall code clarity. Key changes include:

### ✅ Summary of Changes:

* Replaced `phpstan/phpstan` with `larastan/larastan` for better Laravel integration.
* Raised PHPStan level to `4` and updated `phpstan.neon` configuration.
* Improved type annotations throughout:

  * Method docblocks (`@param`, `@return`, `@var`) with explicit types (e.g. `Collection<int, Media>`)
  * Refined PHP type hints in method signatures.
  * Added `@throws` annotations where exceptions are expected.
* Modernized some property and method definitions (e.g., visibility, return types).
* Cleaned up unnecessary `use` statements and improved PSR compliance.

These changes are **non-breaking** and aim to improve maintainability and developer confidence through enhanced static analysis.

Let me know if any adjustments are needed!

anilkumarthakur60 committed:
Update dependencies and improve PHPStan configuration

- Replaced `phpstan/phpstan` with `larastan/larastan` in `composer.json`.
- Updated PHPStan configuration in `phpstan.neon` to include Larastan extensions and increased analysis level to 2.
- Added PHPStan ignore comments in `ImageManipulator.php` for specific lines.
- Refactored return types and method signatures in `Media.php` and `Mediable.php` for better type safety.
- Enhanced `MediableCollection` methods to accept array|string for tags.
- Added new media loading methods in `SampleMediable.php` for better tag handling.

anilkumarthakur60 committed:
Update PHPStan level and improve type annotations in commands and jobs

- Increased PHPStan analysis level from 2 to 3 in `phpstan.neon`.
- Enhanced type annotations in `ImportMediaCommand.php` and `CreateImageVariants.php` for better type safety, specifying `Collection<int, Media>` for existing media collections and models.

anilkumarthakur60 committed:
Update PHPStan configuration and enhance type safety across multiple files

- Increased PHPStan analysis level from 3 to 4 in `phpstan.neon`.
- Added an ignore error rule for unused traits in PHPStan configuration.
- Improved type annotations in `HandlesMediaUploadExceptions.php`, `ImageManipulator.php`, `Media.php`, and `MediaUploader.php` for better type safety.
- Refactored method signatures to include return types and exception documentation.
- Updated visibility modifiers for class constants in `MediaUploader.php` and `StreamAdapter.php`.

anilkumarthakur60 committed:
Update automated test workflow to include Imagick extension for PHP

anilkumarthakur60 committed:
Add format conversion checks and dynamic driver selection

Introduced a driver check to skip tests when format conversion isn't supported under the current configuration. Updated ImageManipulator to dynamically select between 'gd' and 'imagick' drivers based on availability. This ensures compatibility and prevents test failures in unsupported environments.

anilkumarthakur60 committed:
test and ci  reverted

anilkumarthakur60 committed:
fix: minor changes [skip ci]

anilkumarthakur60 committed:
styleci fixes

frasmage reviewed (commented):
This is great! though a few changes are needed

anilkumarthakur60 committed:
fix: update PHPDoc types and improve error handling

anilkumarthakur60 committed:
fix: update PHPDoc types and improve error handling

StyleCIBot committed:
Apply fixes from StyleCI

anilkumarthakur60 committed:
test fixes

anilkumarthakur60 committed:
Merge branch 'phpstan-level-4-fixes' into analysis-W6BQ9y

anilkumarthakur60 committed:
Merge pull request #2 from anilkumarthakur60/analysis-W6BQ9y

Apply fixes from StyleCI

anilkumarthakur60 committed:
fix: update migration class syntax and improve PHPDoc comments

anilkumarthakur60 committed:
dock block fixes

anilkumarthakur60 committed:
commit reverted

anilkumarthakur60 committed:
Refactor code formatting and improve type checks.

Condensed conditional statements for better readability and updated PHPStan annotations for Intervention Image compatibility. Added missing exception annotations to methods and loosened type constraints in `SourceAdapterFactory` for flexibility.

anilkumarthakur60 committed:
Refactor parameter types and clean up formatting

Updated type declarations for better consistency and clarified parameter order in `makeVariantOf`. Simplified `if` condition in `ImageManipulator` and removed redundant newlines in `MediaUploader`. These changes improve readability and maintainability.

frasmage commented on src/Mediable.php:
```diff
@@ -628,7 +630,8 @@ public function load($relations)
         }
 
-        if (array_key_exists('media', $relations)
+        if (
```
style nit: no newline after if

frasmage commented on src/MediableCollection.php:
```diff
@@ -13,9 +14,10 @@
+ * @template TMedia of Model&MediableInterface
  * @extends Collection<TKey, TMedia>
  */
+
```
style nit: no newline after docblock

frasmage commented on src/MediableServiceProvider.php:
```diff
@@ -210,7 +211,8 @@ public function registerConsoleCommands(): void
     {
         $imageManager = null;
-        if ($app->has(ImageManager::class)
+        if (
```
style nit: no newline after if

frasmage commented on src/SourceAdapters/StreamAdapter.php:
```diff
@@ -205,7 +204,8 @@ private function getHttpResponseCode(): ?int
         }
         $headers = $this->originalSource->getMetadata('wrapper_data');
-        if (!empty($headers)
+        if (
```
style nit: no newline after if

frasmage commented on tests/migrations/2016_06_27_000001_create_mediable_test_tables.php:
```diff
@@ -5,7 +5,11 @@
+ * Create a table for a mock mediable model.
+=======
  * Create table for mock mediable model
+>>>>>>> parent of 288a94f (Apply fixes from StyleCI)
```
change: merge conflict

frasmage commented on src/Commands/ImportMediaCommand.php:
```diff
@@ -102,7 +103,7 @@ protected function listFiles(string $disk, string $directory = '', bool $recursi
      * Search through the record list for one matching the provided path.
      * @param  string $path
-     * @param  Collection $existingMedia
+     * @param Collection<string|int, Media> $existingMedia
```
suggest: I'm pretty sure `Collection<Media>` is sufficient. Otherwise, I checked the Laravel source, they use `array-key` as key type, probably best to keep consistent if the key type must be defined

```suggestion
     * @param Collection<array-key, Media> $existingMedia
```

frasmage commented on src/ImageManipulator.php:
```diff
@@ -142,6 +145,7 @@ public function createImageVariant(
             $image = $this->imageManager->read($media->contents());
         } else {
             // Intervention Image <3.0
+            /** @phpstan-ignore-next-line */
```
switch to `@phpstan-ignore <error type>`

frasmage commented on src/ImageManipulator.php:
```diff
@@ -342,15 +352,15 @@ private function getHashFromStream(StreamInterface $stream, string $algo): strin
-            && $variant->getDiskPath() === $originalVariant->getDiskPath()
-        ) {
+    ): void {
+        if ($originalVariant && $variant->disk === $originalVariant->disk && $variant->getDiskPath() === $originalVariant->getDiskPath()) {
```
style nit: please don't change the multiline codestyle

frasmage commented on src/ImageManipulator.php:
```diff
@@ -392,13 +402,17 @@ private function generateUniqueFilename(Media $model): string
-        if (class_exists(StreamCommand::class)) {
+        if (class_exists(Intervention\Image\Commands\StreamCommand::class)) {
             // Intervention Image  <3.0
+            /** @phpstan-ignore-next-line */
```
switch to @phpstan-ignore <error type>

frasmage commented on src/Media.php:
```diff
@@ -562,9 +569,7 @@ protected function getMediaMover(): MediaMover
-        if (static::hasGlobalScope(SoftDeletingScope::class)
-            && (!property_exists($this, 'forceDeleting') || !$this->forceDeleting)
-        ) {
+        if (static::hasGlobalScope(SoftDeletingScope::class) && (!property_exists($this, 'forceDeleting') || !$this->forceDeleting)) {
```
style nit: please don't change the multiple line conditional style

frasmage commented on src/MediaUploader.php:
```diff
@@ -1118,13 +1117,11 @@ public function getOptions(): array
-        if (empty($this->config['image_manipulation'])
-            || $model->aggregate_type !== Media::TYPE_IMAGE
-        ) {
+        if (empty($this->config['image_manipulation'])|| $model->aggregate_type !== Media::TYPE_IMAGE) {
```
style nit: please don't change the multiple line conditional style

frasmage reviewed (commented):
Went through with a fine toothed comb. Just a handful of nits to cleanup (and one merge conflict), then it should be looking good
